core/channel.go

```diff
@@ -112,23 +112,32 @@ type Channel struct {
 }
 
 func (c *Connection) newChannel(reqChanBufSize, replyChanBufSize int) *Channel {
-	// create new channel
-	chID := uint16(atomic.AddUint32(&c.maxChannelID, 1) & 0x7fff)
-	channel := &Channel{
-		id:                  chID,
-		conn:                c,
-		msgCodec:            c.codec,
-		msgIdentifier:       c,
-		reqChan:             make(chan *vppRequest, reqChanBufSize),
-		replyChan:           make(chan *vppReply, replyChanBufSize),
-		replyTimeout:        DefaultReplyTimeout,
-		receiveReplyTimeout: ReplyChannelTimeout,
-	}
+	var channel *Channel
+	// Retry until the channel ID we pick is available
+	for {
+		// create new channel
+		chID := uint16(atomic.AddUint32(&c.maxChannelID, 1) & 0x7fff)
+		channel = &Channel{
+			id:                  chID,
+			conn:                c,
+			msgCodec:            c.codec,
+			msgIdentifier:       c,
+			reqChan:             make(chan *vppRequest, reqChanBufSize),
+			replyChan:           make(chan *vppReply, replyChanBufSize),
+			replyTimeout:        DefaultReplyTimeout,
+			receiveReplyTimeout: ReplyChannelTimeout,
+		}
 
-	// store API channel within the client
-	c.channelsLock.Lock()
-	c.channels[chID] = channel
-	c.channelsLock.Unlock()
+		// store API channel within the client
+		c.channelsLock.Lock()
+		_, found := c.channels[chID]
+		if !found {
+			c.channels[chID] = channel
+			c.channelsLock.Unlock()
+			break
+		}
+		c.channelsLock.Unlock()
+	}
```

Comment:
We throw away a Channel if the id is already taken, which is a bit wasteful.

Maybe we could create a channel first, without setting the id. Then grab the mutex and repeatedly set the id to the next one and break once we got a unique one?

There's no problem in bumping the counter while holding the lock; it's dirt cheap.

Reply:
Thanks, that makes sense. I assume we no longer need the atomic in this case (maxChannelD is not used elsewhere).